tokio/tests/sync_rwlock.rs

```diff
@@ -0,0 +1,163 @@
+use tokio_test::task::spawn;
+use tokio_test::{assert_pending, assert_ready};
+
+// multiple reads should be Ready
+#[test]
+fn read_shared_ready() {
+    let rwlock = RwLock::new(100);
+    let mut t1 = spawn(rwlock.read());
+    assert_ready!(t1.poll());
+    let mut t2 = spawn(rwlock.read());
+    assert_ready!(t2.poll());
+}
+
+// when there is an active shared ownded, exclusive access should not be possible
+#[test]
+fn write_shared_pending() {
+    let rwlock = RwLock::new(100);
+    let mut t1 = spawn(rwlock.read());
+    let _g1 = assert_ready!(t1.poll());
+    let mut t2 = spawn(rwlock.write());
+    assert_pending!(t2.poll());
+}
+
+// Acquire an RwLock nonexclusively by a single task
+#[tokio::test]
+async fn read_uncontested() {
+    let rwlock = RwLock::new(100);
+    let result = *rwlock.read().await;
+
+    assert_eq!(result, 100);
+}
+
+// Acquire an RwLock nonexclusively by two different tasks simultaneously .
+#[tokio::test]
+async fn read_shared() {
+    let rwlock = RwLock::new(100);
+
+    let (tx0, rx0) = oneshot::channel::<()>();
+    let (tx1, rx1) = oneshot::channel::<()>();
+    let task0 = rwlock.read().then(move |guard| {
+        tx1.send(()).unwrap();
+        rx0.map(move |_| *guard)
+    });
+    let task1 = rwlock.read().then(move |guard| {
+        tx0.send(()).unwrap();
+        rx1.map(move |_| *guard)
+    });
+
+    let result = future::join(task0, task1).await;
+
+    assert_eq!(result, (100, 100));
+}
+
+// Attempt to acquire an RwLock for reading that already has a writer
+#[tokio::test]
+async fn read_contested() {
+    let rwlock = RwLock::<u32>::new(0);
+    let (tx0, rx0) = oneshot::channel::<()>();
+
+    let task1 = rwlock.write().then(move |mut guard| {
+        *guard += 5;
+        rx0.map_err(|_| {
+            drop(guard);
+        })
+    });
+
+    let task2 = rwlock.read().map(|guard| *guard);
+    let task3 = rwlock.read().map(|guard| *guard);
+    let task4 = async { tx0.send(()) };
+
+    let result = future::join4(task1, task2, task3, task4).await;
+
+    assert_eq!(result, (Ok(()), 5, 5, Ok(())));
+}
+
+// Attempt to acquire an RwLock for reading after a writer has requested access
+#[tokio::test]
+async fn read_write_contested() {
+    let rwlock = RwLock::<u32>::new(42);
+
+    let task1 = rwlock.read().map(move |guard| *guard);
+    let task2 = rwlock.write().map(|mut guard| *guard += 1);
+    let task3 = rwlock.read().map(|guard| *guard);
+
+    let result = future::join3(task1, task2, task3).await;
+    assert_eq!(result, (42, (), 43));
+
+    let g = rwlock.read().await;
+    assert_eq!(*g, 43);
+}
+
+// Acquire an uncontested RwLock in exclusive mode.  poll immediately returns
+// Poll::Ready
+#[tokio::test]
+async fn write_uncontested() {
+    let rwlock = RwLock::new(100);
+    let mut result = rwlock.write().await;
+    *result += 50;
+    assert_eq!(*result, 150);
+}
+
+// RwLocks should be acquired in the order that their Futures are waited upon.
```

Comment:
I'm not sure what this test is actually testing? How could it be that these pushes _don't_ occur in-order?

Reply:
I think the comment should read "... are waited upon, not the order of their creation" and the code should look like this:
```rust
let fut2 = rwlock.write().map(|mut guard| guard.push(2));
let fut1 = rwlock.write().map(|mut guard| guard.push(1));
fut1.await;
fut2.await;
```
Notice that they're created in one order, and polled in the other.

Reply:
yeah, missed that when transcribed them, thanks @asomers!